modules/core/src/main/java/org/apache/ignite/internal/processors/cache/consistentcut/ConsistentCutManager.java

```diff
@@ -0,0 +1,224 @@
+    }
+
+    /**
+     * Set transaction Consistent Cut ID.
+     *
+     * @param tx Transaction.
+     */
+    public void setTransactionCutId(IgniteInternalTx tx) {
+        ConsistentCutFuture cut = cutFutRef.get();
+
+        if (cut != null)
+            tx.cutId(cut.id());
+    }
+
+    /**
+     * Wraps a transaction message if Consistent Cut is running.
+     *
+     * @param txMsg Transaction message to wrap.
+     * @param txCutId Consistent Cut ID after which transaction committed, if specified.
+     */
+    public static GridCacheMessage wrapMessage(
+        GridCacheSharedContext<?, ?> cctx,
+        GridCacheMessage txMsg,
+        @Nullable UUID txCutId
+    ) {
+        if (cctx.consistentCutMgr() == null)
+            return txMsg;
+
+        ConsistentCutFuture cut = cctx.consistentCutMgr().cutFuture();
+
+        if (cut != null)
+            return new ConsistentCutAwareMessage(txMsg, cut.id(), txCutId);
+
+        return txMsg;
+    }
+
+    /**
+     * Handles received Consistent Cut ID from remote node. It compares it with the latest ID that local node is aware of.
+     * Init local Consistent Cut procedure if received ID is a new one.
+     *
+     * @param id ID of {@link ConsistentCutFuture}.
+     */
+    public void handleConsistentCutId(UUID id) {
+        ConsistentCutFuture cutFut = cutFutRef.get();
+
+        if (cutFut != null || Objects.equals(id, lastFinishedCutId))
+            return;
+
+        ConsistentCutFuture newCutFut = clientNode() ? new ClientConsistentCutFuture(id) : newBaselineConsistentCut(id);
+
+        if (!cutFutRef.compareAndSet(cutFut, newCutFut))
+            return;
+
+        if (newCutFut.isDone())
+            return;
+
+        cctx.kernalContext().pools().getSnapshotExecutorService().submit(() -> {
+            BaselineConsistentCutFuture cut = (BaselineConsistentCutFuture)newCutFut;
+
+            try {
+                cut.init();
+
+                if (log.isDebugEnabled())
+                    log.debug("Prepared Consistent Cut: " + id);
+            }
+            catch (IgniteCheckedException e) {
+                U.error(log, "Failed to handle Consistent Cut: " + id, e);
+
+                cut.onDone(e);
+            }
+        });
+    }
+
+    /** Creates new Consistent Cut future for baseline nodes. */
+    protected BaselineConsistentCutFuture newBaselineConsistentCut(UUID id) {
+        return new BaselineConsistentCutFuture(cctx, id);
+    }
+
+    /**
+     * Cancels local Consistent Cut with error.
+     *
+     * @param err Error.
+     */
+    public void cancelCut(Throwable err) {
+        ConsistentCutFuture cut = cutFutRef.get();
+
+        if (cut != null && !cut.isDone())
+            ((BaselineConsistentCutFuture)cut).onDone(err);
+
+        cleanLocalCut();
+    }
+
+    /**
+     * @return Current running Consistent Cut future, if cut isn't running then {@code null}.
+     */
+    public @Nullable ConsistentCutFuture cutFuture() {
```

Comment:
cutFuture, cutFutRef but all variables named cut.
Can we make naming consistent?